cpp/src/io/text/multibyte_split.cu

```diff
@@ -0,0 +1,401 @@
+}
+
+std::unique_ptr<column> create_char_column(rmm::device_uvector<char>&& values)
+{
+  auto size  = values.size();
+  auto dtype = cudf::data_type{type_id::INT8};
+
+  return std::make_unique<cudf::column>(dtype, size, values.release(), rmm::device_buffer(), 0);
+}
+
+std::unique_ptr<column> create_strings_column(rmm::device_uvector<char>&& chars,
+                                              rmm::device_uvector<int32_t>&& offsets,
+                                              rmm::cuda_stream_view stream,
+                                              rmm::mr::device_memory_resource* mr)
+{
+  auto num_strings    = offsets.size() - 1;
+  auto chars_column   = create_char_column(std::move(chars));
+  auto offsets_column = create_column(std::move(offsets));
+
+  return cudf::make_strings_column(
+    num_strings, std::move(offsets_column), std::move(chars_column), 0, {}, stream, mr);
+}
+
+void fork_stream(std::vector<rmm::cuda_stream_view> streams, rmm::cuda_stream_view stream)
+{
+  cudaEvent_t event;
+  cudaEventCreate(&event);
+  cudaEventRecord(event, stream);
+  for (uint32_t i = 0; i < streams.size(); i++) {
+    cudaStreamWaitEvent(streams[i], event, 0);
+  }
+  cudaEventDestroy(event);
+}
+
+void join_stream(std::vector<rmm::cuda_stream_view> streams, rmm::cuda_stream_view stream)
+{
+  cudaEvent_t event;
+  cudaEventCreate(&event);
+  for (uint32_t i = 0; i < streams.size(); i++) {
+    cudaEventRecord(event, streams[i]);
+    cudaStreamWaitEvent(stream, event, 0);
+  }
+  cudaEventDestroy(event);
+}
+
+std::vector<rmm::cuda_stream_view> get_streams(int32_t count, rmm::cuda_stream_pool& stream_pool)
+{
+  auto streams = std::vector<rmm::cuda_stream_view>();
+  for (int32_t i = 0; i < count; i++) {
+    streams.emplace_back(stream_pool.get_stream());
+  }
+  return streams;
+}
+
+cudf::size_type multibyte_split_scan_full_source(cudf::io::text::data_chunk_source const& source,
+                                                 cudf::io::text::detail::trie const& trie,
+                                                 scan_tile_state<multistate>& tile_multistates,
+                                                 scan_tile_state<uint32_t>& tile_offsets,
+                                                 device_span<cudf::size_type> output_buffer,
+                                                 device_span<char> output_char_buffer,
+                                                 rmm::cuda_stream_view stream,
+                                                 std::vector<rmm::cuda_stream_view> const& streams)
+{
+  CUDF_FUNC_RANGE();
+  cudf::size_type chunk_offset = 0;
+
+  multibyte_split_init_kernel<<<TILES_PER_CHUNK, THREADS_PER_TILE, 0, stream.value()>>>(  //
+    -TILES_PER_CHUNK,
+    TILES_PER_CHUNK,
+    tile_multistates,
+    tile_offsets,
+    cudf::io::text::detail::scan_tile_status::oob);
+
+  auto multistate_seed = multistate();
+  multistate_seed.enqueue(0, 0);  // this represents the first state in the pattern.
```

Comment:
seems like `multistate` is always created with first state, maybe there should be way to create a `multistate`, including the state.